universe/common/unity/options_auto.go

```diff
@@ -0,0 +1,190 @@
+
+const (
+	AttributeSlotKindInvalid AttributeSlotKind = ""
+	AttributeSlotKindTexture AttributeSlotKind = "texture"
+	AttributeSlotKindVideo   AttributeSlotKind = "video"
+	AttributeSlotKindNumber  AttributeSlotKind = "number"
+	AttributeSlotKindString  AttributeSlotKind = "string"
+	AttributeSlotKindImage   AttributeSlotKind = "image"
+)
+
+func GetOptionAutoOption(options *entry.AttributeOptions) (*entry.UnityAutoAttributeOption, error) {
+	if options == nil {
+		return nil, nil
+	}
+
+	autoOptionsValue, ok := (*options)["unity_auto"]
+	if !ok {
+		return nil, nil
+	}
+
+	var autoOption *entry.UnityAutoAttributeOption
+	if err := utils.MapDecode(autoOptionsValue, autoOption); err != nil {
+		return nil, errors.WithMessage(err, "failed to decode auto option")
+	}
+
+	return autoOption, nil
+}
+
+func GetOptionAutoMessage(option *entry.UnityAutoAttributeOption, changeType universe.AttributeChangeType,
+	attributeID entry.AttributeID, value *entry.AttributeValue) (*websocket.PreparedMessage, error) {
+
+	// do checks if obligatory fields are present
+	if option == nil {
+		return nil, nil
+	}
+	if option.SlotKind == "" || option.ContentType == "" {
+		return nil, nil
+	}
+
+	data := &AttributeValueChangedMessage{
+		Type: changeType,
+		Data: AttributeValueChangedMessageData{
+			AttributeName: attributeID.Name,
+			Value:         value,
+		},
+	}
+
+	// do some checks depending on slot kind
+	// for textures we need to render them depending on the
+	// content type: currently there are 2 options:
+	// video and text/number/string
+	switch option.SlotKind {
+	case "texture":
+		if option.ContentType == "video" {
+			payload, err := json.Marshal(map[string]interface{}{
+				"url": option.ValueField,
+			})
+			if err != nil {
+				return nil, errors.WithMessage(err, "Failed to marshal preRenderHash")
+			}
+
+			hash, err := renderVideo(payload)
+			if err != nil {
+				return nil, err
+			}
+			if hash != nil {
+				data.Data.Value = hash.Hash
+			}
+		}
+
+		// if we need to render a string or a number we extract the
+		// "string" field of the ValueField
+		start := strings.Index(option.ValueField, "string")
+		end := strings.Index(option.ValueField[start:], ",")
+
+		preRenderHash := option.ValueField[start : start+end]
+
+		payload, err := json.Marshal(map[string]interface{}{
+			"hash": preRenderHash,
+		})
+		if err != nil {
+			return nil, errors.WithMessage(err, "Failed to marshal preRenderHash")
+		}
+
+		hash, err := renderFrame(payload)
+		if err != nil {
+			return nil, err
+		}
+		if hash != nil {
+			data.Data.Value = hash.Hash
+		}
+	}
+
+	sendData, err := json.Marshal(data)
+	if err != nil {
+		return nil, errors.WithMessagef(err, "failed to marshal message payload")
+	}
+
+	topic := string(option.ContentType)
+	if topic == "" {
+		topic = attributeID.PluginID.String()
+	}
+
+	return posbus.NewRelayToUnityMsg(topic, sendData).WebsocketMessage(), errors.Errorf("got invalid slot kind: %s", option.SlotKind)
```

Comment:
I'm not sure, is it designed to send message and return an error?
looks a bit weird